### ISSUE TITLE
Update multiple tables in one query

### ISSUE THREAD

**Issue by Smert** [juanluispaz/ts-sql-query#82]
Hi. Thank you for useful library!

I have tables with 1-to-1 or 1-to-many relations. It is comfortable to update a few tables in one query:

```sql
UPDATE
    user,
    post
SET
    user.last_edit = NOW(),
    post.name = 'Test'
WHERE
    user.id = post.user_id
    AND post.id = 5; 
```

Will it be possible to construct this query using ts-sql-query?

**Comment by juanluispaz** [comment]
Hi!

you're welcome 😉

Please, be aware SQL is massive big, and each database implements different things in different ways; that makes it really hard to talk about SQL as a general query language, instead it just a general concept; on this ts-sql-query offers you a lot because you don't need to care about a lot of small details; additionally ts-sql-query already provides a lot of database-specific features that if you try to use them in the wrong database you will get a compilation error.

Said that, update in almost all databases update a single table at a time; some databases offer you to get data from another table to use in the update, and ts-sql-query already supports this feature, but sets columns of two different tables at the same time is not implemented yet in the current version of the library. In PostgreSQL you can use insert, update or delete as with in a query; that would be a way to do it in a single query (ts-sql-query doesn't implement this specific feature yet, but I've been in talks to implement it in some moment in the future)

I don't say that something is impossible because I receive surprises from time to time. I'm not sure what database you are using; seeing the syntax looks like MySQL or MariaDB. Reviewing MySQL and MariaDB documentation, they have the multiple-table update syntax, but not clear to me if you can set the fields of all tables.

Can you share with me some documentation about this and what databases support it, or at least the one you are using?

Meanwhile, as a workaround, you can go down to regular SQL for this single query or make several updates in a transaction. I really try to make it possible to deal with non-already supported syntaxes in the library, but in this case, is nothing in place yet.

Please, share information on this to try to understand what I can do on this specific feature; and please, share as well how important this is for you and how frequently you will use it.

I will be waiting for your answer.

**Comment by Smert** [comment]
Sorry! I really forgot to say what I mean MySQL.

There are two ways to update multiple tables in one query:

```sql
UPDATE post
INNER JOIN user ON user.id = post.user_id
SET
    user.last_edit = NOW(),
    post.name = 'Test'
WHERE
    post.id = 5;
```
http://sqlfiddle.com/#!9/9bae1d/2

```sql
UPDATE
    user,
    post
SET
    user.last_edit = NOW(),
    post.name = 'Test'
WHERE
    user.id = post.user_id
    AND post.id = 5; 
```

---

* Conversation on StackOverflow: https://stackoverflow.com/questions/4361774/mysql-update-multiple-tables-with-one-query
* Mysql docs: https://dev.mysql.com/doc/refman/8.0/en/update.html

  * > For the multiple-table syntax, [UPDATE](https://dev.mysql.com/doc/refman/8.0/en/update.html) updates rows in each table named in table_references that satisfy the conditions.

  * > The preceding example shows an inner join that uses the comma operator, but multiple-table [UPDATE](https://dev.mysql.com/doc/refman/8.0/en/update.html) statements can use any type of join permitted in [SELECT](https://dev.mysql.com/doc/refman/8.0/en/select.html) statements, such as LEFT JOIN.

**Comment by juanluispaz** [comment]
Understood,

To implement this, I cannot use the object representation of the column to set; how should the query in TS look like?

This will not be an easy change because making TS happy and safe for you is always challenging. Please, let me know how important this is to you.

I will be looking for a way to implement this, but it will require time, and I cannot promise to have it soon. It is possible I will implement it, but I usually try to give you proper solutions in a few days, which is not the case due to the challenge.

**Comment by Smert** [comment]
I think it might looks like:

```typescript
await connection.update(tAdCampaign)
  .setColumn(tUser.lastEdit, connection.currentDateTime())
  .setColumn(tPost.name, 'Test')
  .where(tPost.id.equals(5))
  .executeUpdate();
```

Or like `dynamicPick`:

```typescript
const availableFields = {
    user: {
      lastEdit: tUser.lastEdit,
    },
    name: tPost.name
};
await connection.update(tAdCampaign)
  .setColumns(availableFields, {
    user: {
      lastEdit: connection.currentDateTime()
    },
    name: 'Test'
  })
  .where(tPost.id.equals(5))
  .executeUpdate();
```

**Comment by juanluispaz** [comment]
Hi,

I'm working on ways to change the insert/update/delete input object shape and as well limiting its available columns. With shape functionality, there is a straightforward way to make room to update multiple tables in a single query for MariaDB and MySQL.

The code looks like this:

```ts
const shapedUpdateCustomerNameAndCompanyName = {
    id: 12,
    customerFirstName: 'John',
    customerLastName: 'Smith',
    companyName: 'ACME Inc.'
}

const shapedUpdateCustomerNameAndCompanyNameResult = await connection.update(tCustomer)
    .innerJoin(tCompany).on(tCustomer.companyId.equals(tCompany.id))
    .shapedAs({
        customerFirstName: tCustomer.firstName,
        customerLastName: tCustomer.lastName,
        companyName: tCompany.name
    })
    .set(shapedUpdateCustomerNameAndCompanyName)
    .where(tCustomer.id.equals(shapedUpdateCustomerName.id))
    .executeUpdate()

```

And the generated SQL query is as:

```sql
update 
    customer 
    inner join company on customer.company_id = company.id 
set 
    customer.first_name = ?, 
    customer.last_name = ?, 
    company.name = ? 
where 
    customer.id = ?
```

Due to this `shapeAs` functionality, I will not include any reference in the documentation; meanwhile, the basic update implementation is already there; you can use it and provide feedback to me.

I appreciate your feedback; let me know if this works for you.

**Comment by juanluispaz** [comment]
Released ts-sql-query 1.49.0 with the changes commented here.

**Comment by Smert** [comment]
It works. Thank you!

**Comment by juanluispaz** [comment]
Released ts-sql-query 1.52.0 making public and documenting this feature
